api/mfa/mfa.go

```diff
@@ -0,0 +1,118 @@
+/*
+Copyright 2023 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package mfa
+
+import (
+	"bytes"
+	"context"
+	"encoding/base64"
+	"strings"
+
+	"github.com/gogo/protobuf/jsonpb"
+	"github.com/gravitational/trace"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials"
+	"google.golang.org/grpc/metadata"
+
+	"github.com/gravitational/teleport/api/client/proto"
+)
+
+const (
+	mfaResponseToken                 = "mfa_challenge_response"
+	errAdminActionMFARequiredMessage = "admin-level API request requires MFA verification"
+)
+
+// NewErrAdminActionMFARequired returns an error indicating that an admin-level
+// API request failed due to missing or failing MFA verification.
+func NewErrAdminActionMFARequired() error {
+	// Create the error directly so we don't trace this function call.
+	return &trace.AccessDeniedError{Message: errAdminActionMFARequiredMessage}
+}
+
+// IsErrAdminActionMFARequired returns whether the given error indicates that an
+// admin-level API request failed due to missing or failing MFA verification.
+func IsErrAdminActionMFARequired(err error) bool {
+	return trace.IsAccessDenied(err) && strings.Contains(err.Error(), errAdminActionMFARequiredMessage)
+}
```

Comment:
I'm curious why you chose this direction. If you go with a global var like before you can just use errors.Is for equality (and save us 2 methods).

```suggestion
const mfaResponseToken                 = "mfa_challenge_response"

// ErrAdminActionMFARequired indicates that an admin-level API request failed
// due to a missing MFA payload.
var ErrAdminActionMFARequired = &trace.AccessDeniedError{Message: "admin-level API request requires MFA verification"}
```

Reply:
I had trouble getting this to work as expected. I think the issues is that `trail.FromGRPC` always mutates the error, even if the error was already derived froma gRPC error. So if the Error interceptor runs before this, it causes `error.Is` to incorrectly think it's a different error. Seems like we need to fix `trail.FromGRPC`, but I'l try to get it to work as is.

Reply:
The issue was that I was trying to aggregate the `mfa.CredentialsFromContext` error from the server. This could situationally work given the issue above. Instead, I reverted to the comment that says we should just log that error instead.

Reply:
I had this in mind, that's why I thought it would work: https://github.com/gravitational/trace/blob/6b0832f593fea9869ccb4b2d74a8534caba20164/errors.go#L303-L311C2.

I've fixed some trace issues recently, let me know if I can help. Alternatively, does it work with a plain Go error?